Datastore/tests/System/FilterTest.php

```diff
@@ -0,0 +1,224 @@
+        $data = self::getInitialData();
+        $entities = [];
+        foreach ($data as $element) {
+            $key = self::$restClient->key(self::$kind, $element['Name']);
+            $entities[] = self::$restClient->entity($key, $element);
+        }
+        self::$restClient->insertBatch($entities);
+
+        // on rare occasions the queries below are returning no results when
+        // triggered immediately after an insert operation. the sleep here
+        // is intended to help alleviate this issue.
+        sleep(1);
+
+        foreach ($entities as $entity) {
+            self::$localDeletionQueue->add($entity->key());
+        }
+    }
+
+    public static function tear_down_after_class()
+    {
+        self::tearDownFixtures();
+    }
+
+    /**
+     * @dataProvider defaultDbClientProvider
+     */
+    public function testOrRunQuery(DatastoreClient $client)
+    {
+        $this->skipEmulatorTests();
+        $filter = Filter::doOr([
+            Filter::where('Name', '=', 'Hersch'),
+            Filter::where('Name', '=', 'Davy')
+        ]);
+        $query = $client->query()
+            ->kind(self::$kind)
+            ->filter($filter);
+
+        $results = $this->runQueryAndSortResults($client, $query);
+        $this->assertEquals(count($results), 2);
+        $this->assertEquals($results[0]['Name'], 'Hersch');
+        $this->assertEquals($results[1]['Name'], 'Davy');
+    }
+
+    /**
+     * @dataProvider defaultDbClientProvider
+     */
+    public function testMixOfOrAndRunQuery(DatastoreClient $client)
+    {
+        $this->skipEmulatorTests();
+        $filter = Filter::doAnd([
+            Filter::where('Age', '<', 26),
+            Filter::doOr([
+                Filter::doAnd([
+                    Filter::where('Age', '>', 23),
+                    Filter::where('Age', '<', 30)
+                ]),
+                Filter::doAnd([
+                    Filter::where('Age', '>', 25),
+                    Filter::where('Age', '<', 31)
+                ]),
+            ])
+        ]);
+        $query = $client->query()
+            ->kind(self::$kind)
+            ->filter($filter);
+
+        $results = iterator_to_array($client->runQuery($query));
+        $this->assertEquals(count($results), 1);
+        $this->assertEquals($results[0]['Name'], 'Eldredge');
+    }
+
+    /**
+     * @dataProvider defaultDbClientProvider
+     */
+    public function testOrQueryViaTransaction(DatastoreClient $client)
+    {
+        $this->skipEmulatorTests();
+        $filter = Filter::doOr([
+            Filter::where('Name', '=', 'Hersch'),
+            Filter::where('Name', '=', 'Davy')
+        ]);
+        $query = $client->query()
+            ->kind(self::$kind)
+            ->filter($filter);
+
+        $transaction = $client->transaction();
+        $results = iterator_to_array($transaction->runQuery($query));
+        usort($results, function ($a, $b) {
+            return $a['Name'] < $b['Name'];
+        });
```

Comment:
you can use `$query->order(..)` rather than having this loop

Reply:
also, consider re-using `runQueryAndSortResults`

Reply:
Reused `runQueryAndSortResults ` in recent commits.